hie-plugin-api/Haskell/Ide/Engine/Ghc.hs

```diff
@@ -21,41 +21,49 @@ import           Bag
 import           Control.Monad.IO.Class
 import           Data.IORef
 import qualified Data.Map.Strict                   as Map
+import qualified Data.IntMap.Strict                   as IM
 import           Data.Semigroup ((<>), Semigroup)
 import qualified Data.Set                          as Set
 import qualified Data.Text                         as T
 import qualified Data.Aeson
 import           Data.Coerce
 import           ErrUtils
 
-import qualified GhcModCore                   as GM ( withDynFlags
-                                                    , gcatches, GHandler(..), ghcExceptionDoc
-                                                    , mkErrStyle', renderGm
-                                                    , getModulesGhc'
-                                                    , GmlT(..), getMMappedFiles, GmState(..), GhcModT, cradle
-                                                    , cabalResolvedComponents
-                                                    , IOish, GhcModError(..), GmGhcSession(..), GhcModState(..), GmModuleGraph(..), Cradle(..), gmcHomeModuleGraph
-                                                    , mkRevRedirMapFunc )
-
 import           Haskell.Ide.Engine.MonadFunctions
 import           Haskell.Ide.Engine.MonadTypes
 import           Haskell.Ide.Engine.PluginUtils
-import           System.FilePath
 
 import           DynFlags
+import qualified EnumSet as ES
 import           GHC
 import           IOEnv                             as G
 import           HscTypes
 import           Outputable                        (renderWithStyle)
 import           Language.Haskell.LSP.Types        ( NormalizedUri(..), toNormalizedUri )
 
--- ---------------------------------------------------------------------
+import           Haskell.Ide.Engine.GhcUtils
+--import qualified Haskell.Ide.Engine.Plugin.HieExtras as Hie
+
+import           Outputable hiding ((<>))
+-- This function should be defined in HIE probably, nothing in particular
+-- to do with BIOS
+import qualified HIE.Bios.GHCApi as BIOS (withDynFlags, CradleError,setDeferTypeErrors)
+import qualified HIE.Bios as BIOS
+import Debug.Trace
+
+import System.Directory
+
+import GhcProject.Types as GM
+import Digraph (Node(..), verticesG)
+import GhcMake ( moduleGraphNodes )
+import GhcMonad
+
 
 newtype Diagnostics = Diagnostics (Map.Map NormalizedUri (Set.Set Diagnostic))
   deriving (Show, Eq)
 
 instance Semigroup Diagnostics where
-  Diagnostics d1 <> Diagnostics d2 = Diagnostics (d1 <> d2)
+  Diagnostics d1 <> Diagnostics d2 = Diagnostics (Map.unionWith Set.union d1 d2)
```

Comment:
@wz1000 made this change, I wonder why?

Reply:
Probably as otherwise some errors are discarded because of the `Monoid` semantics for `Map`.